### PR TITLE
[build-script] Refactor toolchain

### PR DESCRIPTION
**Pull request by rintaro**
#### What's in this pull request?

As a ground work for [SR-237](https://bugs.swift.org/browse/SR-237), and regarding https://github.com/apple/swift/pull/2295#discussion_r61532199
Refactored `toolchain` in `build-script`.
Provides consistent cross platform API to retrieve host tool paths.

``` python
from swift_build_support.toolchain import host_toolchain

# `toolchain` always exists, and represents host toolchain
toolchain = host_toolchain()

# tools path can be accessed as properties
cc = toolchain.cc  # /path/to/clang
cxx = toolchain.cxx  # /path/to/clang++
distcc = toolchain.distcc  # /path/to/distcc or None

# toolchain is mutable
toolchain.ninja = '/path/to/my/ninja'
```

Renamed functions:
- `toolchain.host_clang` -> `toolchain.cc`, `toolchain.cxx`
- `cmake.host_cmake` -> `toolchain.cmake`
- `ninja.is_ninja_installed` -> `toolchain.ninja` (is not None)

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

### CODE REVIEW

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by rintaro** [comment]
@harlanhaskins Could you take a look at this PR?

**Comment by harlanhaskins** [comment]
LGTM with regards to the things I've touched. Could @gribozavr take a look and see if I'm missing anything with integrating `toolchain` into the rest of the pipeline?

**Comment by gribozavr** [comment]
@swift-ci Please test

File: utils/swift_build_support/swift_build_support/toolchain.py [hunk 1]
```diff
@@ -11,124 +9,172 @@
...
+class Darwin(Toolchain):
+    def __init__(self, sdk, toolchain):
+        super(Darwin, self).__init__()
+        self.xcrun_sdk = sdk
+        self.xcrun_toolchain = toolchain
+
+    def find_tool(self, *names):
+        for name in names:
+            # NOTE: xcrun searches from developer tools directory *and* from
+            #       PATH. Relatively slow, but we don't need `which `for
```
Comment (gribozavr) [inline utils/swift_build_support/swift_build_support/toolchain.py]:
```
  `which `for  => `which` for
```

**Comment by gribozavr** [comment]
@rintaro LGTM modulo the whitespace comment.

**Comment by rintaro** [comment]
@harlanhaskins @gribozavr Thank you! fixed.

**Comment by gribozavr** [comment]
@swift-ci Please test and merge

**Pull request merged**